src/sentry/issues/occurrence_status_change.py

```diff
@@ -0,0 +1,58 @@
+from __future__ import annotations
+
+import logging
+from dataclasses import dataclass
+from typing import Sequence
+
+from sentry.models.group import Group, GroupStatus
+from sentry.types.activity import ActivityType
+
+logger = logging.getLogger(__name__)
+
+
+@dataclass(frozen=True)
+class OccurrenceStatusChange:
+    fingerprint: Sequence[str]
+    project_id: int
+    new_status: int
+    new_substatus: int | None
+
+    def update_status_for_occurrence(self, group):
+        if group.status == self.new_status and group.substatus == self.new_substatus:
+            return
+
+        # Validate the provided status and substatus - we only allow setting a substatus for unresolved or ignored groups.
+        if self.new_status in [GroupStatus.UNRESOLVED, GroupStatus.IGNORED]:
+            if self.new_substatus is None:
+                logger.error(
+                    "group.update_status.missing_substatus",
+                    extra={
```

Comment:
@snigdhas Nit: i'd put these into variables that can be re-used

Reply:
Ah yeah I think I've cleaned tests up in [this PR](https://github.com/getsentry/sentry/pull/58301) - I'll leave the changes there to avoid having to sync them down up the chain